blog/content/post/2020-12-08-IOMMU.md

```diff
@@ -0,0 +1,162 @@
+  contiguous to a device (scatter/gather). Scatter/gather optimizes streaming DMA
+  performance for the I/O device.
+  - Memory isolation and protection: device can only access memory regions that are
+  mapped for it. Hence faulty and/or malicious devices can't corrupt memory.
+  - Memory isolation allows safe device assignment to a virtual machine without
+  compromising host and other guest OSes.
+  - IOMMU enables 32-bit DMA capable devices to access to > 4GB memory.
+  - Support hardware interrupt remapping. It extends limited hardware interrupts
+  to software interrupts. Primary uses are the interrupt isolation and translation
+  between interrupt domains (E.g. ioapic vs x2apic on x86)
+
+Disadvantages of IOMMU usage:
+
+  - Latency in dynamic DMA mapping, translation overhead penalty.
+  - Host software has to maintain in-memory data structures for use by the IOMMU
+
+## In hardware
+
+IOMMU’s architecture is designed to accommodate a variety of system topologies.
+There can be multiple IOMMUs located at a variety of places in the system fabric.
+It can be placed at bridges inside one bus or at bridges between busses of the
+same or different types. In modern systems, IOMMU is commonly integrated with the
+PCIe root complex. It connects to the PCIe bus from downstream and to the system
+bus ( E.g. infinity fabric) from upstream. One example topology is shown below.
+
+![Example Platform Architecture](/img/iommu.png) 
+
+## In software
+
+The IOMMU is configured and controlled via two sets of registers. One in the PCI
+configuration space and another set mapped in system address space. Since the
+IOMMU appears to OS as a PCI function, it has a capability block in the PCI
+configuration space. Additionally, up to eight data structures are placed in the
+main system memory. These data structures contain information about the mapping
+of interrupts and memory for the usage of peripheral devices. Information about
+the device's memory access permissions is stored there too.
+
+## Enabling
+
+IOMMU is a generic name for, be Intel VT-d and AMD-Vi technologies. Make sure
+that your CPU supports one of these before you try to enable IOMMU.
+ 
+### BIOS
+
+First of all, IOMMU has to be initiated by BIOS and information about it has to be
+passed to the kernel in ACPI tables. For the end-user, that means that you have
+to enter the BIOS settings and set IOMMU option to `enabled`. For readers
+interested in developing firmware capable of IOMMU initialization, the next post
+in this topic will describe the process of enabling IOMMU for APU2 in coreboot.
+
+### Linux kernel
+Enable IOMMU support by setting the correct kernel parameter depending on the
+type of CPU in use:
+
+- For Intel CPUs (VT-d) set `intel_iommu=on`
+- For AMD CPUs (AMD-Vi) set `amd_iommu=on`
+- Additionally if you interested in PCIe passthrough set `iommu=pt`
+
+After rebooting, you can check dmesg output to confirm that IOMMU is enabled
+```sh
+dmesg | grep -i -e DMAR -e IOMMU
```

Comment:
@Sakoram this is valid only for Intel, either we should underline that fact either show also AMD side.